cdap/resource_streaming_program_run.go

```diff
@@ -0,0 +1,335 @@
+				Description: "Specifies if multiple runs of the same program should be allowed",
+				DefaultFunc: func() (interface{}, error) {
+					return false, nil
+				},
+			},
+		},
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(20 * time.Minute),
+			Delete: schema.DefaultTimeout(time.Hour), // This gives the pipeline time to drain processing if in-flight records
+		},
+	}
+}
+
+func resourceStreamingProgramRunCreate(d *schema.ResourceData, m interface{}) error {
+	config := m.(*Config)
+
+	addr := getProgramAddr(config, d)
+	startAddr := urlJoin(addr, "start")
+	runsAddr := urlJoin(addr, "runs")
+
+	argsObj := make(map[string]string)
+	// cast map[string]interface{} to map[string]string
+	for k, val := range d.Get("runtime_arguments").(map[string]interface{}) {
+		argsObj[k] = val.(string)
+	}
+
+	runId, _ := uuid.NewRandom()
+	// This runtime arg will be unused by the pipeline but will allow the provider to associate a run with this resource.
+	argsObj[FAUX_RUN_ID] = runId.String()
+
+	b, err := json.Marshal(argsObj)
+	if err != nil {
+		return err
+	}
+
+	req, err := http.NewRequest(http.MethodPost, startAddr, bytes.NewReader(b))
+	if err != nil {
+		return err
+	}
+
+	if _, err := httpCall(config.httpClient, req); err != nil {
+		return err
+	}
+
+	// Poll until actually reaches RUNNING state.
+	return resource.Retry(d.Timeout(schema.TimeoutCreate), func() *resource.RetryError {
+		isRunning, err := isFauxRunIdRunningYet(config, runsAddr, runId.String())
+		if err != nil {
+			return resource.RetryableError(err)
+		}
+		if isRunning {
+			d.SetId(runId.String())
+			return nil
+		}
+		time.Sleep(10 * time.Second) // avoid spamming retries
+		return resource.RetryableError(fmt.Errorf("still waiting for program run with faux run id: %s which is in an initializing state", runId.String()))
+	})
+}
+
+//TODO ?
+func resourceStreamingProgramRunRead(d *schema.ResourceData, m interface{}) error {
+	return nil
+}
+
+type RuntimeArgs struct {
+	FauxRunId string `json:"__FAUX_RUN_ID__"`
+}
+
+type RuntimeProperties struct {
+	RuntimeArgs json.RawMessage `json:"runtimeArgs"`
+}
+
+// These are the only keys we need
+type Run struct {
+	RunId      string `json:"runid"`
+	Status     string `json:"status"`
+	Properties RuntimeProperties
+}
+
+// Checks if there is a running run for the terraform faux run id
+// raises error if the program is not in an initializing state (e.g. it failed or was killed in the ui)
+func isFauxRunIdRunningYet(config *Config, runsAddr string, runId string) (bool, error) {
+	s, err := getProgramRunStatusByFauxId(config, runsAddr, runId)
+	if err != nil {
+		return false, err
+	}
+
+	if s == "RUNNING" {
+		return true, nil
+	}
+	if !INITIALIZING_PROGRAM_RUN_STATUSES[s] {
+		return false, fmt.Errorf("Program not Running or Initializing, in state: %s", s)
+	}
+	return false, nil
+
+}
+
+func getProgramRunStatusByFauxId(config *Config, runsAddr string, runId string) (s string, err error) {
```

Comment:
This func doesn't seem to do much, just get the status in the caller

Reply:
Yeah agreed, I was trying to be DRY. 
We do this twice and can avoid an extra:
    if err != nil {
        return err
    }

    s := r.Status

Reply:
You don't need to do s := r.Status, just using r.Status multiple times is fine